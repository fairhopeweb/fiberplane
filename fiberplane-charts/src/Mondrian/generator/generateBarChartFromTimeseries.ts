import type {
  AbstractChart,
  Axis,
  Shape,
  ShapeList,
  TimeseriesSourceData,
} from "../types";
import {
  calculateBarWidth,
  calculateBarX,
  calculateSmallestTimeInterval,
  calculateYAxisRange,
  createMetricBuckets,
  extendAxisWithInterval,
  extendMinMax,
  getInitialMinMax,
  getTimeFromTimestamp,
  getXAxisFromTimeRange,
  MinMax,
  normalizeAlongLinearAxis,
} from "./utils";
import { compact, identity } from "../../utils";
import type { Metric, Timeseries } from "../../providerTypes";

export function generateBarChartFromTimeseries(
  input: TimeseriesSourceData,
): AbstractChart<Timeseries, Metric> {
  const visibleTimeseriesData = input.timeseriesData.filter(
    (timeseries) => timeseries.visible,
  );

  const buckets = createMetricBuckets(
    visibleTimeseriesData,
    (maybeMinMax: MinMax | undefined, value) =>
      maybeMinMax ? extendMinMax(maybeMinMax, value) : getInitialMinMax(value),
  );

  const xAxis = getXAxisFromTimeRange(input.timeRange);
  const yAxis = calculateYAxisRange(buckets, identity);

  const numShapeLists = visibleTimeseriesData.length;
  const interval = calculateSmallestTimeInterval(buckets);
  if (interval) {
    extendAxisWithInterval(xAxis, interval);
  }

  const barWidth = calculateBarWidth(xAxis, interval ?? 0, numShapeLists);
  const barArgs: BarArgs = { barWidth, numShapeLists, xAxis, yAxis };

  const shapeLists: Array<ShapeList<Timeseries, Metric>> =
    input.timeseriesData.map((timeseries) => ({
      shapes: timeseries.visible
        ? compact(
            timeseries.metrics.map((metric) =>
              getBarShape(
                metric,
                visibleTimeseriesData.indexOf(timeseries),
                barArgs,
              ),
            ),
          )
        : [],
      source: timeseries,
    }));

  return { shapeLists, xAxis, yAxis };
}

type BarArgs = {
  barWidth: number;
  numShapeLists: number;
  xAxis: Axis;
  yAxis: Axis;
};

function getBarShape(
  metric: Metric,
  barIndex: number,
  { xAxis, yAxis, barWidth, numShapeLists: numVisibleTimeseries }: BarArgs,
): Shape<Metric> | null {
  if (Number.isNaN(metric.value)) {
    return null;
  }

  const groupX = normalizeAlongLinearAxis(
    getTimeFromTimestamp(metric.time),
    xAxis,
  );
  return {
    type: "rectangle",
    x: calculateBarX(groupX, barWidth, barIndex, numVisibleTimeseries),
    width: barWidth,
    y: 0,
    height: normalizeAlongLinearAxis(metric.value, yAxis),
    source: metric,
  };
}
