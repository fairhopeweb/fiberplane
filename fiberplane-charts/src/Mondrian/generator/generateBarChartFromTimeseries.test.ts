import type { Metric, Timeseries } from "../../providerTypes";
import { generateBarChartFromTimeseries } from "./generateBarChartFromTimeseries";

test("it generates a bar chart", () => {
  const timeseries: Timeseries = {
    name: "dummy_data",
    labels: {},
    attributes: {},
    resource: {},
    metrics: [
      getMetricAtMinute(0, 10),
      getMetricAtMinute(1, 15),
      getMetricAtMinute(2, 20),
      getMetricAtMinute(4, 25),
      getMetricAtMinute(5, 30),
      getMetricAtMinute(6, 25),
      getMetricAtMinute(7, 20),
      getMetricAtMinute(8, 15),
      getMetricAtMinute(9, 10),
    ],
    visible: true,
  };

  expect(
    generateBarChartFromTimeseries({
      graphType: "bar",
      stackingType: "none",
      timeRange: {
        from: getDateAtMinute(0).toISOString(),
        to: getDateAtMinute(10).toISOString(),
      },
      timeseriesData: [timeseries],
      additionalValues: [],
    }),
  ).toMatchSnapshot();
});

function getMetricAtMinute(minute: number, value: number): Metric {
  return {
    time: getDateAtMinute(minute).toISOString(),
    value,
    attributes: {},
    resource: {},
  };
}

function getDateAtMinute(minute: number): Date {
  return new Date(
    `2023-07-18T16:${minute.toString().padStart(2, "0")}:00.000Z`,
  );
}
