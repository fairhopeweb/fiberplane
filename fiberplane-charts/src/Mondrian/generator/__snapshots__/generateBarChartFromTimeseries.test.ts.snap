// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it generates a bar chart 1`] = `
{
  "shapeLists": [
    {
      "shapes": [
        {
          "height": 0.045454545454545456,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 10,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": -0.0031565656565656535,
          "y": 0,
        },
        {
          "height": 0.2727272727272727,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 15,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.08775252525252525,
          "y": 0,
        },
        {
          "height": 0.5,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.17866161616161616,
          "y": 0,
        },
        {
          "height": 0.7272727272727273,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 25,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.360479797979798,
          "y": 0,
        },
        {
          "height": 0.9545454545454546,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.4513888888888889,
          "y": 0,
        },
        {
          "height": 0.7272727272727273,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 25,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.5422979797979799,
          "y": 0,
        },
        {
          "height": 0.5,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 20,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.6332070707070707,
          "y": 0,
        },
        {
          "height": 0.2727272727272727,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 15,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.7241161616161617,
          "y": 0,
        },
        {
          "height": 0.045454545454545456,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 10,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.8150252525252526,
          "y": 0,
        },
      ],
      "source": {
        "attributes": {},
        "labels": {},
        "metrics": [
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 10,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 15,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 25,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 25,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 15,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 10,
          },
        ],
        "name": "dummy_data",
        "resource": {},
        "visible": true,
      },
    },
  ],
  "xAxis": {
    "maxValue": 1689696630,
    "minValue": 1689695970,
    "tickSuggestions": [
      1689696000,
      1689696060,
      1689696120,
      1689696180,
      1689696240,
      1689696300,
      1689696360,
      1689696420,
      1689696480,
      1689696540,
      1689696600,
    ],
  },
  "yAxis": {
    "maxValue": 31,
    "minValue": 9,
  },
}
`;
